models/AppState.py

```diff
@@ -1,23 +1,148 @@
+            if len(product) == 0:
+                sys.tracebacklimit = 0
+                raise Exception(f'Market not found! ({self.app.getMarket()})')
+
+            base = float(self.account.getBalance(self.app.getBaseCurrency()))
+            base_min = float(product['base_min_size'])
+
+            if base < base_min:
+                sys.tracebacklimit = 0
+                raise Exception(f'Insufficient Base Funds! (Actual: {base}, Minimum: {base_min})')
+
+    def minimumOrderQuote(self):
+        if self.app.getExchange() == 'binance':
+            info = self.api.client.get_symbol_info(symbol=self.app.getMarket())
+
+            quote_min = 0
+            if 'filters' in info:
+                for filter_type in info['filters']:
+                    if 'MIN_NOTIONAL' == filter_type['filterType']:
+                        quote_min = float(filter_type['minNotional'])
+
+            quote = float(self.account.getBalance(self.app.getQuoteCurrency()))
+
+            if quote < quote_min:
+                sys.tracebacklimit = 0
+                raise Exception(f'Insufficient Quote Funds! (Actual: {quote}, Minimum: {quote_min})')
+
+        elif self.app.getExchange() == 'coinbasepro':
+            product = self.api.authAPI('GET', f'products/{self.app.getMarket()}')
+            if len(product) == 0:
+                sys.tracebacklimit = 0
+                raise Exception(f'Market not found! ({self.app.getMarket()})')
+
+            ticker = self.api.authAPI('GET', f'products/{self.app.getMarket()}/ticker')
+            price = float(ticker['price'])
+
+            quote = float(self.account.getBalance(self.app.getQuoteCurrency()))
+            base_min = float(product['base_min_size'])
+
+            if (quote / price) < base_min:
+                sys.tracebacklimit = 0
+                raise Exception(f'Insufficient Quote Funds! (Actual: {"{:.8f}".format((quote / price))}, Minimum: {base_min})')
+
+    @classmethod
+    def initLastAction(cls, app:PyCryptoBot, account:TradingAccount, state):
+        # ignore if manually set
+        if app.getLastAction() is not None:
+            state.last_action = app.getLastAction()
+            return
+
+        orders = account.getOrders(app.getMarket(), '', 'done')
+        if len(orders) > 0:
+            last_order = orders[-1:]
+
+            # if orders exist and last order is a buy
+            if str(last_order.action.values[0]) == 'buy':
+                state.last_buy_size = float(last_order[last_order.action == 'buy']['size'])
+                state.last_buy_filled = float(last_order[last_order.action == 'buy']['filled'])
+                state.last_buy_price = float(last_order[last_order.action == 'buy']['price'])
+                state.last_buy_fee = float(last_order[last_order.action == 'buy']['fees'])
+                state.last_action = 'BUY'
+                return
+            else:
+                cls.minimumOrderBase()
```

Comment:
@iThom @emreeroglu, can you help me with this please? If you pull down this branch you will see it complains about lines 122, 140, and 143.

The error I'm getting:
No value for argument 'self' in unbound method call.      pylint(no-value-for-parameter)

This is the reason I moved those functions outside the class as I couldn't get it working. I moved the functions back within the class. Can you see what is wrong here?

Reply:
I think you are getting this error, because you are calling instance methods from a class method.
Instance method's first parameter is instance's itself, class method's is class' itself.

I think solution would be not using class method here, so every AppState() instance can be capable to initialize and decide its own last_action, this way seems more future proof to me.

When we convert this function to  instance function we will not need state as parameter of function we will only call self.variable to modify instance variables.

Reply:
Do you mind making this update so I can see what you mean?

Reply:
Made the changes but could not run the sim.
(installed the requirements file in the PR)
Got this error:

Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 1683, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'market'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/mnt/d/emre/workspace/projects/personal/cryptobot/.venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 2897, in get_loc
    raise KeyError(key) from err
KeyError: 'market'

Reply:
Thanks, I'll look.